lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -12,27 +12,181 @@ namespace {
+
+  template <unsigned int N> bool numOfChildrenEq(Expr exp) {
+    return exp->arity() == N;
+  }
+
+  template <unsigned int N> bool numOfChildrenGreaterEq(Expr exp) {
+    return exp->arity() >= N;
+  }
+
+  bool isValidNumType(Expr exp) {
+    if (exp != nullptr &&
+        (isOp<INT_TY>(exp) || isOp<REAL_TY>(exp) || isOp<UNINT_TY>(exp)))
+      return true;
+    return false;
+  }
+
+  // ensures the expression has the correct number of children and all children
+  // are the same and correct type
+  Expr numTypeCheckChildren(Expr exp,
+                            std::function<bool(Expr exp)> checkNumChildren) {
+    if (!checkNumChildren(exp))
+      return Expr();
+
+    Expr type = m_cache.at(exp->first());
+    auto isSameType = [this, type](Expr exp) {
+      return type != nullptr && this->m_cache.at(exp) == type;
+    };
+
+    if (isValidNumType(type) &&
+        std::all_of(exp->args_begin(), exp->args_end(), isSameType))
+      return type;
+    else
+      return Expr();
+  }
+
+  Expr inferTypeNumUnary(Expr exp) {
+    std::function<bool(Expr)> checkNumChildren =
+        std::bind(&TCVR::numOfChildrenEq<1>, this, std::placeholders::_1);
+    return numTypeCheckChildren(exp, checkNumChildren);
+  }
+
+  Expr inferTypeNumNary(Expr exp) {
+    std::function<bool(Expr)> checkNumChildren = std::bind(
+        &TCVR::numOfChildrenGreaterEq<2>, this, std::placeholders::_1);
+    return numTypeCheckChildren(exp, checkNumChildren);
+  }
+
+  Expr inferTypeITE(Expr exp) {
+    // ite(a,b,c) : a is bool type, b and c are the same type
+    if (numOfChildrenEq<3>(exp) && isOp<BOOL_TY>(m_cache.at(exp->arg(0))) &&
+        (m_cache.at(exp->arg(1)) == m_cache.at(exp->arg(2))))
+      return sort::boolTy(exp->efac());
+    else
+      return Expr();
+  }
+
+  // ensures the expression has the correct number of children and all children
+  // are bool types
+  Expr boolCheckChildren(Expr exp,
+                         std::function<bool(Expr exp)> checkNumChildren) {
+    auto isBool = [this](Expr exp) {
+      Expr type = this->m_cache.at(exp);
+
+      return type != nullptr && isOp<BOOL_TY>(this->m_cache.at(exp));
+    };
+
+    if (checkNumChildren(exp) &&
+        std::all_of(exp->args_begin(), exp->args_end(), isBool))
+      return sort::boolTy(exp->efac());
+    else
+      return Expr();
+  }
+
+  Expr inferTypeBoolBinary(Expr exp) {
+    std::function<bool(Expr)> checkNumChildren =
+        std::bind(&TCVR::numOfChildrenEq<2>, this, std::placeholders::_1);
+    return boolCheckChildren(exp, checkNumChildren);
+  }
+
+  Expr inferTypeBoolUnary(Expr exp) {
+    std::function<bool(Expr)> checkNumChildren =
+        std::bind(&TCVR::numOfChildrenEq<1>, this, std::placeholders::_1);
+    return boolCheckChildren(exp, checkNumChildren);
+  }
+
+  Expr inferTypeBoolNary(Expr exp) {
+    std::function<bool(Expr)> checkNumChildren = std::bind(
+        &TCVR::numOfChildrenGreaterEq<2>, this, std::placeholders::_1);
```

Comment:
@agurfinkel is this what you meant when you recommended to use templates instead of lambdas?

Reply:
Something got lost in translation. Don't use `std::bind` if you can. Use lambdas.
You had function pointers before (I think), and I was suggesting to use `std::function` for them instead.

Reply:
oh ok I must have misread something